### DISCUSSION TITLE
Guidance on Property Thickness Optimization in TACS and OpenMDAO

### DISCUSSION THREAD

**Discussion by MohdAhmet** [OpenMDAO/mphys#191] [Q&A]
Dears, 
I am working on optimizing a grillage structure's design using TACS and OpenMDAO. My Nastran file is fully developed in FEMAP, and my optimization problem is defined as follows:

- Objective: Minimize structural mass
- Constraints: Yield stress and maximum deflection
- Design Variables: Thickness of specific properties (IDs are identified in the Nastran file).

While exploring the documentation, I noticed most examples focus on optimizing element thickness, which differs from my case. Could you kindly guide me to a relevant example or provide insight on connecting dv_struct to the specific property thicknesses in my model?

And also, can I initialize the tacs builder without element callback function while doing the optimization setup? as I have everything predefined in my bdf file.

Thank you in advance for your support!

**Comment by MohdAhmet** [comment]
I'm sorry to ask to much, but I am really struggling with successfully running my optimization.

###########################################################
import os
import openmdao.api as om
from mphys import Multipoint
from mphys.scenario_structural import ScenarioStructural
from tacs import TACS, functions
from tacs.mphys import TacsBuilder

# Path to your BDF file
bdf_file = os.path.join(os.getcwd(), "m1-005.dat")

# Define initial PSHELL property (thickness for PSHELL ID 1)
pshell_property1 = 11.5e-3  # Thickness for PSHELL ID 1


def problem_setup(scenario_name, fea_assembler, problem):
    """
    Setup the structural problem.
    """
    problem.addFunction("mass", functions.StructuralMass)
    problem.addFunction("ks_stress", functions.KSFailure, safetyFactor=1.5, ksWeight=100.0)
    problem.addFunction("compliance", functions.Compliance)

# Define the multipoint model
class HatchCoverModel(Multipoint):
    def setup(self):
        # TACS setup options passed directly to TacsBuilder
        struct_builder = TacsBuilder(
            mesh_file=bdf_file,
            problem_setup=problem_setup,
            coupled=False,
            check_partials=True
        )
        struct_builder.initialize(self.comm)

        # Add subsystems
        self.add_subsystem("mesh", struct_builder.get_mesh_coordinate_subsystem())
        self.mphys_add_scenario("hatch_cover_analysis", ScenarioStructural(struct_builder=struct_builder))
        self.mphys_connect_scenario_coordinate_source("mesh", "hatch_cover_analysis", "struct")

        # Add independent variable for PSHELL thickness
        dvs = self.add_subsystem("dvs", om.IndepVarComp(), promotes=["*"])
        dvs.add_output("thickness_1", val=pshell_property1)  # Single thickness design variable

        # Connect design variable to the structural model
        self.connect("thickness_1", "hatch_cover_analysis.dv_struct")

# OpenMDAO problem setup
prob = om.Problem()
prob.model = HatchCoverModel()

# Add design variable, constraints, and objectives
prob.model.add_design_var("thickness_1", lower=0.004, upper=0.03)  # Thickness bounds (4mm to 30mm)

prob.model.add_objective("hatch_cover_analysis.mass", scaler=1.0)
prob.model.add_constraint("hatch_cover_analysis.ks_stress", upper=1.0)  # Yield strength constraint
prob.model.add_constraint("hatch_cover_analysis.compliance", upper=0.095)  # Deflection limit

# Configure optimizer
prob.driver = om.ScipyOptimizeDriver(debug_print=["objs", "nl_cons"], maxiter=200)
prob.driver.options["optimizer"] = "SLSQP"
prob.driver.options['maxiter'] = 100  # Set the number of iterations to 100

# Setup and execute
prob.setup()
om.n2(prob, show_browser=False, outfile="hatch_cover_struct.html")
prob.run_driver()

# Output results
if prob.comm.rank == 0:
    thickness_1 = prob.get_val("thickness_1")[0]  # Extract the scalar value from the array
    mass = prob.get_val("hatch_cover_analysis.mass")[0]
    print(f"Optimized Thickness for PSHELL ID 1: {thickness_1:.6f} m")
    print(f"Final Mass: {mass:.6f} kg")

####################################################################################

My script is working, but the result I am getting is only one iteration, no change of thickness and no mass change as well.
I am doubting that my property thickness is not linked to my bdf file because I have't defined element callback function, I haven;t done so because everything is predefined in my nastran file. However, I tried to understand how to define it without affecting me model, but got very confused.

Can someone please give some hints or guidance, on how to optimize one or several properties, that are predefined in my nastran file. 

thanks in advance, sorry again for bothering

**Comment by timryanb** [comment]
Closing since this question is being asked on tacs issue page here: https://github.com/smdogroup/tacs/issues/344
